File: ts/112.path-sum.ts
/*
 * @lc app=leetcode id=112 lang=typescript
 *
 * [112] Path Sum
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// @lc code=start

function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
  if (root === null) return false;

  targetSum -= root.val;

  if (root.left === null && root.right === null) return targetSum === 0;

  return hasPathSum(root.left, targetSum) || hasPathSum(root.right, targetSum);
}

// @lc code=end

File: go/112.path-sum/112.path-sum.go
/*
 * @lc app=leetcode id=112 lang=golang
 *
 * [112] Path Sum
 */

package leetcode

import "github.com/maguroid/leetcode/go/structures"

type TreeNode = structures.TreeNode

// @lc code=start

func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	if root.Left == nil && root.Right == nil {
		return root.Val == targetSum
	}

	return hasPathSum(root.Left, targetSum-root.Val) || hasPathSum(root.Right, targetSum-root.Val)
}

// @lc code=end
